serialization/pickle.ksy

```diff
@@ -0,0 +1,443 @@
+  Other builtin types, and all classes  (e.g. `set`, `datetime.datetime`) are
+  serialised by encoding the name of a constructor callable.
+  They are deserialised by importing that constructor, and calling it.
+doc-ref:  https://github.com/python/cpython/blob/3.3/Lib/pickletools.py
+seq:
+  # TODO is there a way to declare PROTO is optional, but only valid at position 0?
+  - id: ops
+    type: op
+    repeat: eos
+  # TODO is there a way to declare a trailing STOP is required?
+types:
+  op:
+    seq:
+      - id: code
+        type: u1
+        enum: opcode
+        doc: |
+          Operation code that determines which action should be
+          performed next by the Pickel Virtual Machine. Some opcodes
+          are only available in later versions of the Pickle protocol.
+      - id: arg
+        type:
+          switch-on: code
+          cases:
+            'opcode::int': decimalnl_short
+            'opcode::binint': s4
+            'opcode::binint1': u1
+            'opcode::binint2': u2
+            'opcode::long': decimalnl_long
+            'opcode::long1': long1
+            'opcode::string': stringnl
+            'opcode::binstring': string4
+            'opcode::short_binstring': string1
+            'opcode::binbytes': bytes4
+            'opcode::short_binbytes': bytes1
+            #'opcode::none': no_arg
+            #'opcode::newtrue': no_arg
+            #'opcode::newfalse': no_arg
+            'opcode::unicode': unicodestringnl
+            'opcode::binunicode': unicodestring4
+            'opcode::binfloat': f8
+            #'opcode::empty_list': no_arg
+            #'opcode::append': no_arg
+            #'opcode::appends': no_arg
+            #'opcode::list': no_arg
+            #'opcode::empty_tuple': no_arg
+            #'opcode::tuple': no_arg
+            #'opcode::tuple1': no_arg
+            #'opcode::tuple2': no_arg
+            #'opcode::tuple3': no_arg
+            #'opcode::empty_dict': no_arg
+            #'opcode::dict': no_arg
+            #'opcode::setitem': no_arg
+            #'opcode::setitem2': no_arg
+            #'opcode::pop': no_arg
+            #'opcode::dup': no_arg
+            #'opcode::mark': no_arg
+            #'opcode::pop_mark': no_arg
+            'opcode::get': decimalnl_short
+            'opcode::binget': u1
+            'opcode::long_binget': u4
+            'opcode::put': decimalnl_short
+            'opcode::binput': u1
+            'opcode::long_binput': u4
+            'opcode::ext1': u1
+            'opcode::ext2': u2
+            'opcode::ext4': u4
+            'opcode::global':  stringnl_noescape_pair
+            #'opcode::reduce': no_arg
+            #'opcode::build': no_arg
+            'opcode::inst': stringnl_noescape_pair
+            #'opcode::obj': no_arg
+            #'opcode::newobj': no_arg
+            'opcode::proto': u1
+            #'opcode::stop': no_arg
+            'opcode::persid': stringnl_noescape
+            #'opcode::binpersid': no_arg
```

Comment:
All of the `no_arg` cases are commented out here - how are no-arg operations handled then? As far as I can tell, they would currently cause errors, because none of the cases would match.

Reply:
I was unsure of the best way to handle these. Having an empty arg field seemed a bit redundant, on the other hand having a uniform shape for the op type has some appeal. I'm open to either approach.


> As far as I can tell, they would currently cause errors

Not that I've seen so far. I left it this way after getting no errors from the web IDE, when parsing a few samples.

Reply:
>As far as I can tell, they would currently cause errors, because none of the cases would match.

No, it is a completely valid situation. If none of `cases` matches, the field is assigned to `None` and its analogues and consumes `0` space

Reply:
Actually it seems that if no case matches, the field is not parsed at all, which (in Python) results in a missing attribute:

```python
In [1]: !ksc --target=python python_pickle.ksy 

In [2]: import python_pickle                                                                        

In [3]: import pickle                                                                               

In [4]: thing = python_pickle.PythonPickle.from_bytes(pickle.dumps(123))                            

In [5]: [op.code for op in thing.ops]                                                               
Out[5]: [<Opcode.proto: 128>, <Opcode.binint1: 75>, <Opcode.stop: 46>]

In [7]: thing.ops[0].arg                                                                            
Out[7]: 3

In [8]: thing.ops[1].arg                                                                            
Out[8]: 123

In [9]: thing.ops[2].arg                                                                            
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-9-edc885b8c4bf> in <module>
----> 1 thing.ops[2].arg

AttributeError: 'Op' object has no attribute 'arg'
```

I think having explicit `no_arg` cases would be better, so that the code using the spec can always safely access `arg`. You could do this with a default case (`_: no_arg`), but I think listing each no-arg opcode separately would be better, so that you can tell apart opcodes that really have no arguments and ones that just haven't been implemented.